Dicom initial commit

marcelolerendegui opened the pull request:
@yongtang, much appreciated for the help on kicking this off!

marcelolerendegui commented:
> I signed it!

yongtang commented:
@marcelolerendegui  @Ouwen I think we are almost there. Can you update the PR to use the updated `dcmtk.BUILD` from the following (to fix a few macOS issues):

https://raw.githubusercontent.com/tensorflow/io/535b70ea5e23d5e8e4e9131fc700552f41cdea4d/third_party/dcmtk.BUILD

Once dcmtk.BUILD is updated I think we are ready to get this PR merged.

yongtang commented:
@marcelolerendegui @Ouwen  There is another error:
```
E     File "/v/tests/test_dicom_eager.py", line 18
E       from __future__ import absolute_import
E   SyntaxError: from __future__ imports must occur at the beginning of the file
```

Can you also remove the empty line between the comment and the `from import`:
```
"""Tests for DICOM."""

from __future__ import absolute_import
```

marcelolerendegui commented:
@yongtang Seems like nongnu.org is down.

feihugis commented on tensorflow_io/dicom/kernels/decode_dicom_image.cc:
```diff
@@ -0,0 +1,293 @@
+        TensorShape out_shape;
+        if ((samples_per_pixel == 1) && (color_dim == false))
+        {
+            out_shape = TensorShape({static_cast<int64>(frameCount), static_cast<int64>(frameHeight), static_cast<int64>(frameWidth)});
```
The line length needs to be less than 80 chars. You can check the C++ code style using `clang-format`. More details can be found [here](https://github.com/tensorflow/tensorflow/blob/master/CONTRIBUTING.md#c-coding-style).

feihugis commented on tensorflow_io/dicom/kernels/decode_dicom_data.cc:
```diff
@@ -0,0 +1,97 @@
+        DcmDataset *dset = dfile.getDataset();
+        DcmMetaInfo *meta = dfile.getMetaInfo();
+
+        for (unsigned int tag_i = 0; tag_i < in_tags_flat.size(); tag_i++)
```
`++tag_i` may be better.

feihugis commented on tensorflow_io/dicom/kernels/decode_dicom_data.cc:
```diff
@@ -0,0 +1,97 @@
+            DcmTag tag(tag_group_number, tag_element_number);
+
+            OFString val;
+            if (dset->tagExists(tag))
```
The braces had better be added here

feihugis commented on tensorflow_io/dicom/kernels/decode_dicom_image.cc:
```diff
@@ -0,0 +1,293 @@
+            *out_value = (double)(in_value);
+    }
+
+    string on_error;
```
`on_error` -> `on_error_`.

feihugis commented on tensorflow_io/dicom/kernels/decode_dicom_image.cc:
```diff
@@ -0,0 +1,293 @@
+    }
+
+    string on_error;
+    string scale;
```
ditto

feihugis commented on tensorflow_io/dicom/kernels/decode_dicom_image.cc:
```diff
@@ -0,0 +1,293 @@
+
+    string on_error;
+    string scale;
+    bool color_dim;
```
ditto

feihugis commented on tensorflow_io/dicom/kernels/decode_dicom_image.cc:
```diff
@@ -0,0 +1,293 @@
+REGISTER_DECODE_DICOM_IMAGE_CPU(Eigen::half);
+REGISTER_DECODE_DICOM_IMAGE_CPU(double);
+
+#undef REGISTER_DECODE_DICOM_IMAGE_CPU
```
missing new line

feihugis commented on tensorflow_io/dicom/ops/decode_dicom_data_op.cc:
```diff
@@ -0,0 +1,28 @@
+    .Output("tag_values: string")
+    .Doc(R"doc(
+loads a dicom file and returns the specified tags values as string.
+)doc");
```
missing newline

feihugis commented on tensorflow_io/dicom/ops/decode_dicom_image_op.cc:
```diff
@@ -0,0 +1,31 @@
+    .Attr("scale: {'auto', 'preserve'} = 'preserve'")
+    .Doc(R"doc(
+loads a dicom image file and returns its pixel information in the specified output format
+)doc");
```
missing newline

feihugis commented on tensorflow_io/dicom/python/ops/dicom_ops.py:
```diff
@@ -0,0 +1,4076 @@
+  EndingRespiratoryAmplitude = int('00209248', 16)
+  EndingRespiratoryPhase = int('00209249', 16)
+  RespiratoryTriggerType = int('00209250', 16)
+  RRIntervalTimeNo
```
missing newline

feihugis commented on third_party/dcmtk.BUILD:
```diff
@@ -0,0 +1,990 @@
+        "#endif // CONFIG_ARITH_H",
+        "EOF",
+    ]),
+)
```
missing newline

yongtang commented:
All review comments have been addressed and all tests passed, except for the additional lint issue (one with python and another one with bazel):
```
************* Module tests.test_dicom_eager
W: 25, 0: Bad indentation. Found 4 spaces, expected 2 (bad-indentation)
W: 29, 0: Bad indentation. Found 4 spaces, expected 2 (bad-indentation)
W: 31, 0: Bad indentation. Found 4 spaces, expected 2 (bad-indentation)
W: 32, 0: Bad indentation. Found 4 spaces, expected 2 (bad-indentation)
W: 33, 0: Bad indentation. Found 4 spaces, expected 2 (bad-indentation)
W: 37, 0: Bad indentation. Found 4 spaces, expected 2 (bad-indentation)
```

```
--- ./tensorflow_io/dicom/BUILD	2019-07-18 21:52:16.879995001 +0000
+++ /tmp/buildifier-tmp-258320356	2019-07-18 21:52:56.607995001 +0000
@@ -25,4 +25,4 @@
         "@local_config_tf//:libtensorflow_framework",
         "@local_config_tf//:tf_header_lib",
     ],
-)
\ No newline at end of file
+)
exit status 1
```

@marcelolerendegui can you update the PR again?

yongtang commented:
Never mind, I will merge and add a follow up PR for the lint fix. @marcelolerendegui Thanks for the contribution 👍 🎉 🎆 !

marcelolerendegui commented:
Hey @yongtang 

I added sample dicom files and tests.

I found a problem when running the tests: "Warning: no data dictionary loaded"
Aparently DCMTK is looking for the data dictionaries in $DCMDICTPATH.
I changed dcmtk.BUILD to define ENABLE_BUILTIN_DICTIONARY and now it is working.

I downloaded the 21 DCM sample files (totaling 8.3MB) from [here](https://barre.dev/medical/samples/)

Before requesting a pull:
Can these sample files be merged into tensorflow/io?

yongtang commented:
@marcelolerendegui 8.3 MB is a little larger than I expected. Would prefer to download during the test than save it to git which will be in the history.

But if the files are not available through direct download then it should be fine.
